### ISSUE TITLE
Improve support of higher-order derivatives for built-in functions

### ISSUE THREAD

**Issue by kmaehashi** [chainer/chainer#4449]
Although most of built-in functions now support higher-order derivative, some functions still do not support it.

Currently following functions do not support higher-order derivatives:

* ~~`F.slstm`~~ (#4824) merged!
* `F.tree_lstm`
* `F.spatial_transformer_grid`
* `F.spatial_transformer_sampler`
* `F.depthwise_convolution_2d`
* `F.n_step_bigru`
* `F.n_step_bilstm`
* `F.n_step_birnn`
* `F.n_step_gru`
* `F.n_step_lstm`
* `F.n_step_rnn`
* `F.accuracy`
* `F.binary_accuracy`
* `F.classification_summary`
* `F.r2_score`
* ~~`F.contrastive` (#3830)~~ merged!
* `F.connectionist_temporal_classification`
* `F.decov`
* `F.hinge`
* ~~`F.prod`~~ (#3764) merged!
* `F.batch_renormalization`
* `F.local_response_normalization`
* `F.roi_pooling_2d`
* ~~`F.forget`~~ (#3792) merged!
* `TheanoFunction`

The following functions only support second order derivatives:

* `F.lstm`
* `F.batch_normalization`
* `F.fixed_batch_normalization`

**Comment by kmaehashi** [comment]
Guides for people considering contributions to do this work:

* To make function support higher-order derivative (aka double backprop), you need to reimplement `chainer.Function` as `chianer.FunctionNode`.
  [New-Style Function Implementation Guideline](https://gist.github.com/beam2d/dd87d97b8d7581f95cd19b1de86c9f4d) describes the detailed steps to do this work.
* You can find example of PRs in https://github.com/chainer/chainer/issues/3147 and https://github.com/chainer/chainer/issues?q=new+style+-label%3Abackport.
